File: mcp-prompt-health/prompt-quality-server.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import axios from "axios";

// Inisialisasi MCP server
type ToolInput = {
  client_uuid: string;
  project: string;
  efektivitas: number;
  membingungkan: number;
};

const server = new McpServer({
  name: "prompt-quality",
  version: "1.0.0",
});

// Daftarkan tool submit_prompt_quality
server.registerTool(
  "submit_prompt_quality",
  {
    title: "Submit Prompt Quality",
    description: "Kirim skor kualitas prompt ke backend Laravel.",
    inputSchema: {
      client_uuid: z.string().uuid(),
      project: z.string(),
      efektivitas: z.number().min(0).max(100),
      membingungkan: z.number().min(0).max(100),
    },
  },
  async ({ client_uuid, project, efektivitas, membingungkan }: ToolInput) => {
    const payload = {
      client_uuid,
      project,
      prompt_quality: { efektivitas, membingungkan },
    };

    const base = process.env.PROMPT_QUALITY_API ?? "http://localhost:8000";
    const resp = await axios.post(`${base}/api/prompt-quality`, payload, {
      timeout: 10000,
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json",
      },
    });

    return {
      content: [
        {
          type: "text",
          text: `✅ Terkirim (HTTP ${resp.status})`,
        },
      ],
    } as const;
  }
);

// Jalankan server MCP via STDIO
async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

main().catch((err) => {
  console.error("Failed to start MCP server:", err);
  process.exit(1);
});

// Jalankan dengan:
//   npx tsx --esm prompt-quality-server.ts
// Atau jika menggunakan CommonJS mode:
//   npx tsx prompt-quality-server.ts  (tanpa top-level await di file ini)

File: mcp-prompt-health/prompt_quality_server.py
"""Prompt-quality MCP server.

This module exposes a single MCP tool named ``submit_prompt_quality`` that will be
consumed by Cursor (or any other MCP-compatible client).  The tool receives a
prompt-quality score and forwards it to a Laravel backend.  The implementation
is intentionally kept minimal, but we still try to:

* Remain compatible with both *fastmcp* ≥ 2.9 (published as ``fastmcp``) and the
  legacy package distributed under ``mcp.server.fastmcp``.
* Fail fast with a helpful log message if either dependency is missing or if
  the outbound HTTP request fails.
"""

from __future__ import annotations

import asyncio
import logging
import os
from typing import Any

import httpx

# ---------------------------------------------------------------------------
# FastMCP import compatibility shim
# ---------------------------------------------------------------------------

try:
    # Preferred location since FastMCP ≥ 2.9
    from fastmcp import FastMCP  # type: ignore
except ModuleNotFoundError:  # pragma: no cover — runtime fallback
    # Legacy fallback that works for FastMCP ≤ 2.8.x
    from mcp.server.fastmcp import FastMCP  # type: ignore

# ---------------------------------------------------------------------------
# Logging setup
# ---------------------------------------------------------------------------

logging.basicConfig(
    level=os.getenv("LOGLEVEL", "INFO").upper(),
    format="[%(asctime)s] %(levelname)s %(name)s: %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# MCP server & tool definition
# ---------------------------------------------------------------------------

mcp = FastMCP("prompt-quality")

# ▸ gunakan tanda kurung jika FastMCP ≥2.9 (lihat traceback Anda sebelumnya)
@mcp.tool()
async def submit_prompt_quality(
    project: str,
    efektivitas: int,
    membingungkan: int,
    ambiguous: int = 0,
    comments: str = "",
) -> str:
    """Kirim skor kualitas prompt ke endpoint Laravel.
    
    Args:
        project: Nama project
        efektivitas: Skor efektivitas (1-100)
        membingungkan: Skor membingungkan (1-100)
        ambiguous: Skor ambiguitas (1-100, optional)
        comments: Komentar tambahan (optional)
    """
    # Allow the caller to omit ``client_uuid`` and fall back to the environment
    # variable ``CLIENT_UUID`` (or ``default-uuid`` if not set).
    client_uuid = os.getenv("CLIENT_UUID")

    payload: dict[str, Any] = {
        "client_uuid": client_uuid,
        "project": project,
        "prompt_quality": {
            "efektivitas": efektivitas,
            "membingungkan": membingungkan,
            "ambiguous": ambiguous if ambiguous > 0 else None,
            "comments": comments if comments else None,
        },
    }

    base_url = os.getenv("PROMPT_QUALITY_API", "http://localhost:8000").rstrip("/")
    url = f"{base_url}/api/prompt-quality"

    logger.debug("POST %s • payload=%s", url, payload)

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                url,
                json=payload,
                headers={
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
            resp.raise_for_status()
    except httpx.HTTPError as exc:
        logger.error("Failed to POST to %s: %s", url, exc)
        raise

    logger.info("Prompt quality sent successfully • HTTP %s", resp.status_code)
    return f"✅ terkirim (HTTP {resp.status_code})"

if __name__ == "__main__":
    # Jalankan server MCP via STDIO — persis yang Cursor harapkan
    # The server is entirely asynchronous, but FastMCP provides a synchronous
    # ``run`` helper.  If you prefer to manage the event loop yourself, uncomment
    # the snippet below.

    # asyncio.run(mcp.serve_stdio())

    mcp.run()

